public/app/features/trails/Integrations/interpolation.ts

```diff
@@ -0,0 +1,56 @@
+import { AdHocVariableFilter, DataSourceInstanceSettings } from '@grafana/data';
+import { PromQuery, PrometheusDatasource } from '@grafana/prometheus';
+import { getDataSourceSrv } from '@grafana/runtime';
+import { AdHocFiltersVariable, SceneVariable } from '@grafana/scenes';
+import { DataQuery, DataSourceRef } from '@grafana/schema';
+import { DashboardModel } from 'app/features/dashboard/state';
+import { DashboardScene } from 'app/features/dashboard-scene/scene/DashboardScene';
+
+export function interpolateVariables(
+  dashboard: DashboardModel | DashboardScene,
+  dsInstanceSettings: DataSourceInstanceSettings,
+  queries: DataQuery[]
+) {
+  const promDataSourceForUtility = new PrometheusDatasource({ ...dsInstanceSettings, jsonData: {} });
```

Comment:
why new up a new data source?

Reply:
you can get the query data source using `await getDataSourceSrv().get(query.datasource);`

Reply:
For some reason, I had it in my head that I was still in the 'thread' that was generating the menu items and assumes "await" was off limits. Thanks for reminding me that I am within the callback post user selection, so I can "await" all I like.

Reply:
Hmm... wait, I am still in the thread that generates the menu items.
This means that my labels for the menu items will have to use the variable names, but the execution of the action can take advantage of the interpolation... Unless I keep this weird strategy of constructing the datasource just in time... Let me see...

Reply:
@torkelo 
> why new up a new data source?

Though it is awkward, it does allow me to take advantage of the utility functions in the prom datasource for interpolating variables. By recreating the datasource, I can call them within the context of generating these menu items, without `await`, and use the interpolated queries to parse for metrics/filters to generate labelled menu items and corresponding click handlers.

Without interpolating in advance:
- I cannot label the menu items with the interpolated values
- In certain cases I cannot properly identify prometheus metrics and filters due to the language parser not being to handle non-interpolated variable notation

I tried a few things, but the menu list that we populate doesn't tolerate changes. I don't expect there would be a way to make it optionally operate on `Promise<MenuItem>`, or some kind of refresh callback.

Reply:
I could make this function async without any changes (in scenes version) as addDataTrailPanelAction was called from an async function. 

newing up PrometheusDataSource could be problematic (esp as the team wants to externalize this data source) so it would be loaded async as any plugin.